Automattic-Tracks-iOS/Event Logging/EventLoggingDelegate.swift

```diff
@@ -0,0 +1,33 @@
+import Foundation
+
+public protocol EventLoggingDelegate {
+    /// The event logging system will call this delegate property prior to attempting to upload, giving the application a chance to determine
+    /// whether or not the upload should proceed. If this is not overridden, the default is `false`.
+    var shouldUploadLogFiles: Bool { get }
+
+    /// The event logging system will call this delegate method each time a log file starts uploading.
+    func didStartUploadingLog(_ log: LogFile)
+
+    /// The event logging system will call this delegate method if a log file upload is cancelled by the delegate.
+    func uploadCancelledByDelegate(_ log: LogFile)
+
+    /// The event logging system will call this delegate method if a log file fails to upload.
+    func uploadFailed(withError: Error, forLog: LogFile)
+
+    /// The event logging system will call this delegate method each time a log file finishes uploading.
+    func didFinishUploadingLog(_ log: LogFile)
```

Comment:
We may want to add a comment explaining that these delegates will be called before the actual callback goes through in `EventLoggingUploadManager`. I am not too familiar with the naming and conventions in iOS (anymore) so I'd make sure that the ordering of these callbacks and their names match the SDK to avoid confusion.

Reply:
Good call on documenting it. I think we might be hard pressed for a definitive answer on which it should be, but my thinking was to fire the delegate method as soon as we know whether the request was successful or not (ie – `didFinish` is called as soon as it finishes).

I'm open to discussion on whether this is the best way, but IMHO the delegate should have the most up-to-date info since it's event-based and it's ok for the callback to lag behind – there's no expectation that'll be called instantly following a result, just that it _will_ be called in the future.

Docs updated in 5e34b4ef2d29f14a3c57d82c51a5e406c72edff1.

Reply:
If there isn't a rule, what I'd think about in this case is "Does the delegate require any information from the execution of the callback?" if so, do the `delegate` after.

For example, let's say that `LogUploadCallback` we pass is marking the state of the log file as `uploaded` and the delegate is reading that state in `didFinishUploadingLog`. In that case, we'd want the `callback` to be called before the `delegate`.

I guess maybe we should look at why this confusion exists in the first place and whether we can address that instead. It's hard to suggest an approach here because I don't know exactly how this class is intended to be used, but an important question to that end would be "Why do we need a delegate and a callback at the same time"? Is there a design that gets rid of one? If not, can we shift things around to make it more obvious?

Similar to many of my comments before, making an architectural change here may not bring any meaningful impact for the particular PR at hand and I never try to force it. However, I do like bringing these up because I hope that it'll have an impact on how the developer will think this type of problem in the future and maybe come up with a solution that's clearer in future PRs. That's just my long way of saying, if you don't think it's worth discussing this, I am more than happy to drop it :)

Reply:
I like your point about

> Does the delegate require any information from the execution of the callback?

In this case, it does not – they're completely independent (but complimentary) tools.

In terms of 

> Why do we need a delegate and a callback at the same time?

The callback will be used to say "ok, uploading this file is finished, it's ok to remove this file from the upload queue", while the delegate allows the implementing application to respond to these fine-grained lifecycle events, if appropriate (consider a UI showing a list of files in the queue that need to be uploaded – it could be the delegate, and take appropriate action to update the UI based on what's happening).

Really appreciate your feedback :)

Reply:
In that case, I think this design is totally fine, thanks for explaining all that!